src/jit/flowgraph.cpp

```diff
@@ -25718,3 +25720,330 @@ bool Compiler::fgUseThrowHelperBlocks()
+    {
+        JITDUMP("Method does not have multiple noreturn calls.\n");
+        return;
+    }
+
+    // This transformation requires block pred lists to be built
+    // so that flow can be safely updated.
+    assert(fgComputePredsDone);
+
+    struct ThrowHelper
+    {
+        BasicBlock*  m_block;
+        GenTreeCall* m_call;
+
+        ThrowHelper() : m_block(nullptr), m_call(nullptr)
+        {
+        }
+
+        ThrowHelper(BasicBlock* block, GenTreeCall* call) : m_block(block), m_call(call)
+        {
+        }
+
+        static bool Equals(const ThrowHelper x, const ThrowHelper& y)
+        {
+            return BasicBlock::sameEHRegion(x.m_block, y.m_block) && GenTreeCall::Equals(x.m_call, y.m_call);
+        }
+
+        static unsigned GetHashCode(const ThrowHelper& x)
+        {
+            return static_cast<unsigned>(reinterpret_cast<uintptr_t>(x.m_call->gtCallMethHnd));
+        }
+    };
+
+    typedef JitHashTable<ThrowHelper, ThrowHelper, BasicBlock*> CallToBlockMap;
+
+    CompAllocator   allocator(getAllocator(CMK_TailMergeThrows));
+    CallToBlockMap  callMap(allocator);
+    BlockToBlockMap blockMap(allocator);
+
+    // We will run two passes here.
+    //
+    // The first pass finds candidate blocks and builds the mapping
+    // from candidates to canonical blocks, the second pass modifies flow.
+    int numCandidates = 0;
+
+    // First pass
+    //
+    // Scan for THROW blocks. Note early on in compilation (before morph)
+    // noretrurn blcoks are not marked as BBJ_THROW.
+    //
+    // Walk blocks from last to first so that any branches we
+    // introduce to the canonical blocks end up lexically forward
+    // and there is less jumbled flow to sort out later.
+    for (BasicBlock* block = fgLastBB; block != nullptr; block = block->bbPrev)
+    {
+        // For throw helpers the block should have exactly one statement....
+        // (this isn't guaranteed, but seems likely)
+        Statement* stmt = block->firstStmt();
+
+        if ((stmt == nullptr) || (stmt->GetNextStmt() != nullptr))
+        {
+            continue;
+        }
+
+        // ...that is a call
+        GenTree* const tree = stmt->GetRootNode();
+
+        if (!tree->IsCall())
+        {
+            continue;
+        }
+
+        // ...that does not return
+        GenTreeCall* const call = tree->AsCall();
+
+        if (!call->IsNoReturn())
+        {
+            continue;
+        }
+
+        // Sanity check -- only user funcs should be marked do not return
+        assert(call->gtCallType == CT_USER_FUNC);
+
+        // Ok, we've found a suitable call. See if this is one we know
+        // about already, or something new.
+        BasicBlock* canonicalBlock = nullptr;
+
+        JITDUMP("\n*** Does not return call\n");
+        DISPTREE(call);
+
+        // Have we found an equivalent call already?
+        ThrowHelper key(block, call);
+        if (callMap.Lookup(key, &canonicalBlock))
```

Comment:
Do you need a check that insures that both blocks are in the same try region?

Reply:
The check is there, it's up in the `ThrowHelpers::Equals` method.